File: pairs/kyberslash/x86/detection_curve.c
#include <stdio.h>
#include <stdint.h>
#include <math.h>
#include <x86intrin.h>
/* Detection curve for the KyberSlash division on x86, drift-robust. For each
   amplification factor we time a latency chain of AMP idivs just below the divisor
   boundary (quotient 0) and just above it (quotient 1), INTERLEAVED as back-to-back
   pairs so unpinnable turbo drift cancels within each pair, and report the paired
   t-statistic over the differences. If a per-idiv operand-dependent step existed it
   would grow with AMP; on a constant-time divider the paired t stays low and flat and
   never approaches the leak band, so the x86 null is not a hidden sub-noise step
   waiting for more gain: the nonce pairs cross the leak band at amplification 40
   (t = 15420), the division does not at amplification 32. */
static inline uint64_t rd(void){ unsigned a; return __rdtscp(&a); }
static uint64_t timed(uint32_t base, int amp){
  uint32_t d = base; uint64_t t0 = rd();
  for(int i=0;i<amp;i++){ uint32_t q;
    __asm__ volatile("movl %1,%%eax\n\t xorl %%edx,%%edx\n\t divl %2\n\t movl %%eax,%0"
                     :"=r"(q):"r"(d),"r"(3329u):"eax","edx","cc");
    d = base + (q & 1u); }   /* stays base or base+1: same side of the boundary */
  return (rd() - t0) + (d & 0u);
}
#define M 200000
#define PERMS 2000
static double paired_t(const double *d, int n){
  double mean=0; for(int i=0;i<n;i++) mean+=d[i]; mean/=n;
  double var=0; for(int i=0;i<n;i++) var+=(d[i]-mean)*(d[i]-mean); var/=(n-1);
  return var>0 ? fabs(mean)/sqrt(var/n) : 0.0;
}
int main(void){
  static double diff[M];
  static double perm[M];
  int amps[]={1,2,4,8,16,32};
  printf("RESULT boundary 3329\n");
  for(int k=0;k<6;k++){ int amp=amps[k];
    for(int i=0;i<3000;i++){ (void)timed(3328,amp); (void)timed(3330,amp); }  /* warm */
    /* Randomise which operand is timed first. Measuring the low operand first every
     * time confounds the operand with its position in the pair: any order effect, from
     * cache or pipeline state, lands entirely in one class and a paired test then
     * rejects on the order rather than on the divider. With the order coin-flipped and
     * the difference always taken as low minus high, an order effect averages out and
     * what survives is operand dependence. */
    uint64_t ost = 0xD1B54A32D192ED03ull ^ (uint64_t) amp;
    for(int i=0;i<M;i++){ _mm_lfence();
      ost ^= ost << 13; ost ^= ost >> 7; ost ^= ost << 17;
      double lo, hi;
      if (ost & 1ull) { lo=(double)timed(3328,amp); hi=(double)timed(3330,amp); }
      else            { hi=(double)timed(3330,amp); lo=(double)timed(3328,amp); }
      diff[i]=lo-hi; }
    double t = paired_t(diff, M);
    /* Exact null for THIS design. The measurement is a paired difference of two
     * operands timed back to back, so under the null that the two sides are
     * exchangeable each difference's sign is equally likely either way. Flipping
     * signs is therefore the permutation the design licenses, and it needs no band
     * calibrated on some other harness: the null is built from these samples. */
    unsigned long ge = 0;
    uint64_t st = 0x9E3779B97F4A7C15ull ^ (uint64_t) amp;   /* fixed seed: reproducible */
    for (int b = 0; b < PERMS; b++) {
      double m = 0.0, v = 0.0;
      for (int i = 0; i < M; i++) {
        st ^= st << 13; st ^= st >> 7; st ^= st << 17;      /* xorshift64 */
        perm[i] = (st & 1ull) ? diff[i] : -diff[i];
        m += perm[i];
      }
      m /= M;
      for (int i = 0; i < M; i++) { double e = perm[i] - m; v += e * e; }
      v /= (M - 1);
      double tb = v > 0 ? fabs(m) / sqrt(v / M) : 0.0;
      if (tb >= t) ge++;
    }
    double pval = (double) (1 + ge) / (double) (PERMS + 1);
    /* The effect size matters more than the statistic here: a per-division step would
     * accumulate with amp, so print the mean paired difference both per measurement
     * and divided by amp. A constant per-measurement offset shows up as a per-division
     * figure that FALLS as 1/amp, which is how a placement artifact is told apart from
     * an operand-dependent divider. */
    double mdiff = 0.0; for (int i = 0; i < M; i++) mdiff += diff[i]; mdiff /= M;
    printf("RESULT amp_%d %.4f p %.5f mean %.6f perdiv %.6f\n",
           amp, t, pval, mdiff, mdiff / amp);
  }
  return 0;
}

File: pairs/kyberslash/x86/detection_curve.py
#!/usr/bin/env python3
"""Detection curve for the KyberSlash division on the x86 host. Regenerable here."""
import json, re, subprocess, sys, os
from pathlib import Path
HERE = Path(__file__).resolve().parent
REPO = HERE.parent.parent.parent
exe = HERE / "detection_curve"
# M must match detection_curve.c's per-class measurement count, and the band is the
# same calibrated null the scoring adapter uses, so the figure and the verdict rule
# cannot drift apart.
M = int(re.search(r"#define\s+M\s+(\d+)", (HERE / "detection_curve.c").read_text()).group(1))
PERMS = int(re.search(r"#define\s+PERMS\s+(\d+)",
                      (HERE / "detection_curve.c").read_text()).group(1))
subprocess.run(["gcc", "-O2", str(HERE / "detection_curve.c"), "-o", str(exe), "-lm"], check=True)

# R independent repetitions of the whole curve. One run of this experiment is not a
# result: repeated on the same host it gives |t| = 36 at one amplification and |t| = 0.3
# at the same amplification minutes later, with the sign of the mean difference flipping.
# A single run therefore reports whichever way the noise fell. What is reportable is the
# distribution: if a per-division step were real, the sign of the mean difference would
# agree across runs and the per-division magnitude would hold as amplification grows.
R = int(os.environ.get("CURVE_RUNS", "5"))
runs = []
for r in range(R):
    out = subprocess.run(["taskset", "-c", "2", str(exe)], check=True,
                         capture_output=True, text=True).stdout
    one = {}
    for line in out.splitlines():
        m = re.match(r"RESULT amp_(\d+) (\S+) p (\S+) mean (\S+) perdiv (\S+)", line)
        if m:
            one[int(m.group(1))] = {"abs_t": float(m.group(2)), "p": float(m.group(3)),
                                    "mean_ticks": float(m.group(4)),
                                    "ticks_per_division": float(m.group(5))}
    runs.append(one)

amps = sorted(runs[0])
def agg(a, key):
    return [runs[r][a][key] for r in range(R)]

curve = []
for a in amps:
    ts, ps, ms = agg(a, "abs_t"), agg(a, "p"), agg(a, "mean_ticks")
    pos = sum(1 for m in ms if m > 0)
    curve.append({
        "amp": a,
        "abs_t_runs": [round(x, 3) for x in ts],
        "abs_t_min": round(min(ts), 3), "abs_t_max": round(max(ts), 3),
        "p_runs": ps,
        "runs_significant_raw": sum(1 for x in ps if x <= 0.05),
        "mean_ticks_runs": [round(x, 5) for x in ms],
        "mean_sign_positive_runs": pos,
        "sign_consistent": pos == R or pos == 0,
        "ticks_per_division_runs": [round(x, 6) for x in agg(a, "ticks_per_division")],
    })

sign_consistent = [c["amp"] for c in curve if c["sign_consistent"]]
always_sig = [c["amp"] for c in curve if c["runs_significant_raw"] == R]
doc = {
    "finding": "kyberslash-x86-detection-curve",
    "measured_utc": os.environ.get("MEASURE_UTC", ""),
    "host": json.loads((REPO / "results" / "host.json").read_text()),
    "regenerable": "on this x86 host: taskset -c 2 python3 pairs/kyberslash/x86/detection_curve.py",
    "design": ("Paired: a below-boundary and an above-boundary dividend timed back to "
               "back, which of the two goes first decided by a coin flip per iteration "
               "so measurement order cannot masquerade as operand dependence. Each "
               "measurement is a latency chain of AMP divisions. The per-run decision is "
               "a SIGN-FLIP permutation null over the paired differences, the exact null "
               "this design licenses; no band calibrated on another harness is used. "
               "%d measurements per class per run, %d repetitions of the whole curve."
               % (M, R)),
    "why_repetitions": ("Because one run of this is not reproducible. The same "
                        "amplification gives a large statistic in one run and nothing in "
                        "the next, so a single run reports the noise rather than the "
                        "divider. Only quantities stable across repetitions are read as "
                        "findings here."),
    "measurements_per_class": M,
    "runs": R,
    "permutations_per_point": PERMS,
    "curve": curve,
    "amps_with_consistent_sign": sign_consistent,
    "amps_significant_in_every_run": always_sig,
    "reading": ("A per-division operand-dependent step would show a mean difference of "
                "one sign in every run and a per-division magnitude that does not shrink "
                "as the chain lengthens. Read the sign-consistency and the "
                "ticks_per_division columns, not any single |t|."),
}
(REPO / "results" / "kyberslash_detection_curve.json").write_text(json.dumps(doc, indent=2) + "\n")
for c in curve:
    print(f"amp={c['amp']:3d} |t| {c['abs_t_min']:8.3f}..{c['abs_t_max']:8.3f}  "
          f"sig {c['runs_significant_raw']}/{R}  sign+ {c['mean_sign_positive_runs']}/{R}  "
          f"consistent={c['sign_consistent']}")
print("sign-consistent amps:", sign_consistent)
print("significant in every run:", always_sig)
